libnetdata/json/json.c

```diff
@@ -0,0 +1,294 @@
+
+			break;
+
+		case JSON_BOOLEAN:
+			buffer_strcat(wb, e->data.boolean?"TRUE":"FALSE");
+			break;
+
+		case JSON_NULL:
+			buffer_strcat(wb,"NULL");
+			break;
+	}
+    info("JSON: %s", buffer_tostring(wb));
+	buffer_free(wb);
+	return 0;
+}
+
+size_t json_walk_string(char *js, jsmntok_t *t, size_t start, JSON_ENTRY *e)
+{
+	char old = js[t[start].end];
+	js[t[start].end] = '\0';
+	e->original_string = &js[t[start].start];
+
+	e->type = JSON_STRING;
+	e->data.string = e->original_string;
+	if(e->callback_function) e->callback_function(e);
+	js[t[start].end] = old;
+	return 1;
+}
+
+size_t json_walk_primitive(char *js, jsmntok_t *t, size_t start, JSON_ENTRY *e)
+{
+	char old = js[t[start].end];
+	js[t[start].end] = '\0';
+	e->original_string = &js[t[start].start];
+
+	switch(e->original_string[0]) {
+		case '0': case '1': case '2': case '3': case '4': case '5': case '6': case '7':
+		case '8': case '9': case '-': case '.':
+			e->type = JSON_NUMBER;
+			e->data.number = strtold(e->original_string, NULL);
+			break;
+
+		case 't': case 'T':
+			e->type = JSON_BOOLEAN;
+			e->data.boolean = 1;
+			break;
+
+		case 'f': case 'F':
+			e->type = JSON_BOOLEAN;
+			e->data.boolean = 0;
+			break;
+
+		case 'n': case 'N':
+		default:
+			e->type = JSON_NULL;
+			break;
+	}
+	if(e->callback_function) e->callback_function(e);
+	js[t[start].end] = old;
+	return 1;
+}
+
+size_t json_walk_array(char *js, jsmntok_t *t, size_t nest, size_t start, JSON_ENTRY *e)
+{
+	JSON_ENTRY ne;
+
+	char old = js[t[start].end];
+	js[t[start].end] = '\0';
+	ne.original_string = &js[t[start].start];
+
+	memcpy(&ne, e, sizeof(JSON_ENTRY));
+	ne.type = JSON_ARRAY;
+	ne.data.items = t[start].size;
+	ne.callback_function = NULL;
+	strcpy(ne.name, "");
+	strcpy(ne.fullname, ne.name);
+	if(e->callback_function) e->callback_function(&ne);
+	js[t[start].end] = old;
+
+	size_t i, init = start, size = t[start].size;
+
+	start++;
+	for(i = 0; i < size ; i++) {
+		ne.pos = i;
+		if (!e->name || !e->fullname || strlen(e->name) > JSON_NAME_LEN  - 24 || strlen(e->fullname) > JSON_FULLNAME_LEN -24) {
+		    info("JSON: JSON walk_array ignoring element with name:%s fullname:%s",e->name, e->fullname);
+		    continue;
+		}
+		sprintf(ne.name, "%s[%lu]", e->name, i);
+		sprintf(ne.fullname, "%s[%lu]", e->fullname, i);
+
+		switch(t[start].type) {
+			case JSMN_PRIMITIVE:
+				start += json_walk_primitive(js, t, start, &ne);
+				break;
+
+			case JSMN_OBJECT:
+				start += json_walk_object(js, t, nest + 1, start, &ne);
+				break;
+
+			case JSMN_ARRAY:
+				start += json_walk_array(js, t, nest + 1, start, &ne);
+				break;
+
+			case JSMN_STRING:
+				start += json_walk_string(js, t, start, &ne);
+				break;
+		}
+	}
+	return start - init;
+}
+
+size_t json_walk_object(char *js, jsmntok_t *t, size_t nest, size_t start, JSON_ENTRY *e)
+{
+	JSON_ENTRY ne;
+
+	char old = js[t[start].end];
+	js[t[start].end] = '\0';
+	ne.original_string = &js[t[start].start];
+	strcpy(ne.name, "");
```

Comment:
`ne.name[0] = '\0'` has the same effect and it is way faster.

Reply:
Done